src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,325 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+                    //    "@odata.context": "http://localhost:11001/convention/$metadata#Companies/$delta",
+                    //    "value": [
+                    //        {
+                    //            "@odata.type": "#Namespace.Company",
+                    //            "Id": 1,
+                    //            "Name": "Company02",
+                    //            "MyOverdueOrders@odata.delta": [
+                    //                {
+                    //                    "Id": 1,
+                    //                    "Name": "Order 1",
+                    //                    "Quantity": 9
+                    //                }
+                    //                {
+                    //                    "@odata.id": "Employees(2)/NewFriends(2)/Namespace.MyNewFriend/MyNewOrders(2)",
+                    //                    "Quantity": 9
+                    //                }
+                    //            ]
+                    //        }
+                    //    ]
+                    //}
+
+                    // If we have a request payload above and we are handling the changed values in MyOverdueOrders,
+                    // The apiHandlerOfT is MyOverdueOrdersAPIHandler.
+                    // The object with id 1 will use MyOverdueOrdersAPIHandler since odata path will be MyOverdueOrders(1)
+                    // The object with odata id Employees(2)/NewFriends(2)/Microsoft.Test.E2E.AspNet.OData.BulkOperation.MyNewFriend/MyNewOrders(2)
+                    // will use NewOrdersAPIHandler.
+                    // The codebelow ensures we use the correct handler.
+
+                    IODataAPIHandler odataPathApiHandler = apiHandlerFactory.GetHandler(changedObj.ODataPath);
+
+                    if (odataPathApiHandler != null && changedObj.ODataPath.Any() && apiHandler.ToString() != odataPathApiHandler.ToString())
+                    {
+                        handler = odataPathApiHandler as ODataAPIHandler<TStructuralType>;
+                    }
+                }
+
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                Dictionary<string, object> keyValues = new Dictionary<string, object>();
+
+                foreach (string key in _keys)
+                {
+                    object value;
+
+                    if (changedObj.TryGetPropertyValue(key, out value))                   
+                    {
+                        keyValues.Add(key, value);
+                    }
+                }
+
+                try
+                {
+                    ODataAPIResponseStatus odataAPIResponseStatus = handler.TryGet(changedObj.ODataPath.GetKeys(), out original, out getErrorMessage);
```

Comment:
We should do a check on what the `GetKeys()` method returns. If it doesn't return keys then an exception should be thrown.